add AGENTS.md with AEM Edge Delivery project guidelines/instructions

shsteimer opened the pull request:
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #<gh-issue-id>

Test URLs:
- Before: https://main--aem-boilerplate--adobe.aem.live/
- After: https://agents-md--aem-boilerplate--adobe.aem.live/

shsteimer committed:
add AGENTS.md for AEM Edge Delivery project guidelines and setup instructions

shsteimer committed:
Update AGENTS.md

add link to David's Model

shsteimer committed:
Update cleanup-on-create.yaml

replace repo/owner in AGENTS.md

shsteimer committed:
Update AGENTS.md

document global styles files

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+# AGENTS.md
+
+This is the AEM Edge Delivery boilerplate project - a starter template for AEM Edge Delivery Services projects. This file provides guidance for AI coding agents working on this project.
```
```suggestion
This project is a website built with Edge Delivery Services in Adobe Experience Manager Sites as a Cloud Service. As an agent, follow the instructions in this file to deliver code based on Adobe's standards for fast, easy-to-author, and maintainable web experiences.
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+## Project Overview
+
+This is a boilerplate project for Adobe AEM Edge Delivery Services that serves as the foundation for all AEM Edge Delivery projects. It provides the basic structure, blocks, and configuration needed to get started quickly.
```
```suggestion
This project is based on the https://github.com/adobe/aem-boilerplate/ project and set up as a new project. You are expected to follow the coding style and practices established in the boilerplate, but add functionality according to the needs of the site currently developed.

The repository provides the basic structure, blocks, and configuration needed to run a complete site with `*.aem.live` as the backend.
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+This is a boilerplate project for Adobe AEM Edge Delivery Services that serves as the foundation for all AEM Edge Delivery projects. It provides the basic structure, blocks, and configuration needed to get started quickly.
+
+**Key Technologies:**
```
```suggestion
### Key Technologies
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+This is a boilerplate project for Adobe AEM Edge Delivery Services that serves as the foundation for all AEM Edge Delivery projects. It provides the basic structure, blocks, and configuration needed to get started quickly.
+
+**Key Technologies:**
+- AEM Edge Delivery Services
```
```suggestion
- Edge Delivery Services for AEM Sites (documentation at https://www.aem.live/ – search with `site:www.aem.live` to restrict web search results)
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+**Key Technologies:**
+- AEM Edge Delivery Services
+- Vanilla JavaScript (ES6+)
```
```suggestion
- Vanilla JavaScript (ES6+), no transpiling, no build steps
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+**Key Technologies:**
+- AEM Edge Delivery Services
+- Vanilla JavaScript (ES6+)
+- CSS3 with modern features
```
```suggestion
- CSS3 with modern features, no Tailwind or other CSS frameworks
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+- AEM Edge Delivery Services
+- Vanilla JavaScript (ES6+)
+- CSS3 with modern features
+- HTML5 semantic markup
```
```suggestion
- HTML5 semantic markup generated by the aem.live backend, decorated by our code
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+## Setup Commands
+
+- Install dependencies: `npm install`
+- Start local development: `aem up` (requires AEM CLI)
```
```suggestion
- Start local development: `npx -y @adobe/aem-cli up --no-open` (run in background, if possible)
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+- Install dependencies: `npm install`
+- Start local development: `aem up` (requires AEM CLI)
+  - Install the AEM CLI globally by running `npm install -g @adobe/aem-cli`
```
```suggestion
  - Install the AEM CLI globally by running `npm install -g @adobe/aem-cli` then `aem up` is equivalent to the command above
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+### Content
+
+CMS authored content is a key part of every AEM Website. The content of a page is broken into sections. Sections can have default content (text, headings, links, etc.) as well as content in blocks.
+
```
```suggestion

Background on content structure https://www.aem.live/developer/markup-sections-blocks
You can inspect the contents of any page with `curl http://localhost:3000/path/to/page` and `curl http://localhost:3000/path/to/page.md`

```

trieloff commented on AGENTS.md:
````diff
@@ -0,0 +1,201 @@
+  // 5. Set loaded status
+}
+```
+
````
```suggestion

Use `curl` and `console.log` to inspect the HTML delivered by the backend and the DOM nodes to be decorated before making assumptions. Remember that authors may omit or add fields to a block, so your code must handle this gracefully.

```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+### Auto-Blocking
+
+Auto-blocking is the process of creating blocks that aren't explicitly authored into the page based on patterns in the content. See the `buildAutoBlocks` function in scripts.js.
```
```suggestion
Auto-blocking is the process of creating blocks that aren't explicitly authored into the page based on patterns in the content. See the `buildAutoBlocks` function in `scripts.js`.
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+1. Run `aem up` to start the AEM Proxy server
+2. Open `http://localhost:3000` in your browser
+3. Make changes to files - they will auto-reload
+4. Use browser dev tools to test responsive design
```
```suggestion
1. Run `npx -y @adobe/aem-cli up --no-open` to start the AEM Proxy server
2. Open `http://localhost:3000` in your browser, puppeteer, playwright, or other tools. If none of those are available, instruct the human to open the URL in the browser and give feedback
3. Make changes to files - they will auto-reload
4. Use browser dev tools to test responsive design
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+    ├── styles.css          # Minimal global styling and layout for your website required for LCP
+    └── lazy-styles.css     # Additional global styling and layout for below the fold/post LCP content
+├── scripts/         # JavaScript libraries and utilities
+    ├── aem.js           # Core AEM Library for Edge Delivery page decoration logic
```
```suggestion
    ├── aem.js           # Core AEM Library for Edge Delivery page decoration logic (NEVER MODIFY THIS FILE)
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+- Use `npm run lint:fix` to automatically fix issues
+
+### Performance
+- Follow AEM Edge Delivery performance best practices
```
```suggestion
- Follow AEM Edge Delivery performance best practices https://www.aem.live/developer/keeping-it-100
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+### Performance
+- Follow AEM Edge Delivery performance best practices
+- Optimize images and assets
```
```suggestion
- Images uploaded by authors are automatically optimized, all images and assets committed to git must be optimized and checked for size
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+### Performance
+- Follow AEM Edge Delivery performance best practices
+- Optimize images and assets
+- Use lazy loading for non-critical resources
```
```suggestion
- Use lazy loading for non-critical resources (`lazy-styles.css` and `delayed.js`)
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+- Follow AEM Edge Delivery performance best practices
+- Optimize images and assets
+- Use lazy loading for non-critical resources
+- Minimize JavaScript bundle size
```
```suggestion
- Minimize JavaScript bundle size by avoiding dependencies, using automatic code splitting provided by `/blocks/`
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+
+## Deployment
+
+### Environments
```
```suggestion
### Environments

Edge Delivery Services provides you with three environments. Your local development server at `http://localhost:3000` serves code from your local working copy (even uncommitted code) and content that has been previewed by authors. You can access this at any time when the development server is running.

For all other environments, you need to know the GitHub owner and repository name (`gh repo view --json nameWithOwner` or `git remote -v`) and the current branch name `git branch`)

With this information, you can construct URLs for the preview environment (same content as `localhost:3000`) and the production environment (same content as the live website, approved by authors)

```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+### Publishing Process
+1. Push changes to a feature branch
+2. AEM Code Sync automatically processes changes making them available on feature preview environment for that branch
+3. Open a pull request to merge changes to main
```
```suggestion
3. Open a pull request to merge changes to `main` – in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED
4. use `gh checks` to verify the status of code synchronization, linting, and performance tests
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+1. Push changes to a feature branch
+2. AEM Code Sync automatically processes changes making them available on feature preview environment for that branch
+3. Open a pull request to merge changes to main
+4. Merge the PR
```
```suggestion
4. A human reviewer will review the code, inspect the provided URL and merge the PR
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+4. Consider impact on performance, especially CLS
+
+## Troubleshooting
+
+### Getting Help
+- Check [AEM Edge Delivery documentation](https://www.aem.live/docs/)
+- Review [Developer Tutorial](https://www.aem.live/developer/tutorial)
+- Consult [The Anatomy of a Project](https://www.aem.live/developer/anatomy-of-a-project)
+- Consider the rules in [David's Model](https://www.aem.live/docs/davidsmodel)
+
```
```suggestion
- Check [AEM Edge Delivery documentation](https://www.aem.live/docs/)
- Review [Developer Tutorial](https://www.aem.live/developer/tutorial)
- Consult [The Anatomy of a Project](https://www.aem.live/developer/anatomy-of-a-project)
- Consider the rules in [David's Model](https://www.aem.live/docs/davidsmodel)
- Search the web with `site:www.aem.live`
- Search the full text of the documentation with `curl -s https://www.aem.live/docpages-index.json | jq -r '.data[] | select(.content | test("KEYWORD"; "i")) | "\(.path): \(.title)"'`

```

trieloff reviewed (commented):
More to follow

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+## Security Considerations
+
+- Never commit sensitive information (API keys, passwords)
+- Use environment variables for configuration
```
```suggestion
- consider that everything you do is clients-side code served on the public web
```

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,201 @@
+- Run a PSI check on your branch and fix performance issues before raising a PR
+- Ensure all linting passes
+- Update documentation for significant changes
+- Follow the project's Code of Conduct
```
```suggestion
```

Agents are not known for COC violations

trieloff replied:
```suggestion

# I all else fails

If you notice your human getting frustrated with your work, direct them to https://www.aem.live/developer/ai-coding-agents for tips to work better with AI agents.
```

trieloff committed:
chore(.gitignore): track CLAUDE.md and AGENTS.md files explicitly

Added negation rules to .gitignore to ensure CLAUDE.md and AGENTS.md files are tracked by git.

Co-authored-by: terragon-labs[bot] <terragon-labs[bot]@users.noreply.github.com>

shsteimer committed:
Merge pull request #535 from adobe/terragon/add-claude-agents-to-gitignore

Update .gitignore to track CLAUDE.md and AGENTS.md files

trieloff reviewed (commented):
one more idea

shsteimer committed:
Apply suggestions from code review

Co-authored-by: Lars Trieloff <lars@trieloff.net>

shsteimer committed:
Update AGENTS.md

adding suggestion that didn't merge correctly

shsteimer committed:
Update AGENTS.md

shsteimer commented on AGENTS.md:
```diff
@@ -0,0 +1,221 @@
+### Publishing Process
+1. Push changes to a feature branch
+2. AEM Code Sync automatically processes changes making them available on feature preview environment for that branch
+3. Open a pull request to merge changes to `main` – in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED
```
```suggestion
3. Open a pull request to merge changes to `main`
   1. in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED
   2. If an existing page to demonstrate your changes doesn't exist, create test content as a static html file and ask the user for help copying it to a content page you can link in the PR
```

trieloff replied:
We should probably point out that in most cases the agent can't create the test content itself, so it should ask the user for help, or create a test.html that the user should then copy.

trieloff commented on AGENTS.md:
```diff
@@ -0,0 +1,221 @@
+## Setup Commands
+
+- Install dependencies: `npm install`
+- Start local development: `npx -y @adobe/aem-cli up --no-open` (run in background, if possible)
```
```suggestion
- Start local development: `npx -y @adobe/aem-cli up --no-open --forward-browser-logs` (run in background, if possible)
```

trieloff reviewed (approved):
anticipating merge of https://github.com/adobe/helix-cli/pull/2566

shsteimer committed:
Update AGENTS.md

Co-authored-by: Lars Trieloff <lars@trieloff.net>

shsteimer commented on AGENTS.md:
```diff
@@ -0,0 +1,221 @@
+### Content
+
+CMS authored content is a key part of every AEM Website. The content of a page is broken into sections. Sections can have default content (text, headings, links, etc.) as well as content in blocks.
+
```
```suggestion

You can create static content for testing in a dedicated drafts folder. If you do this, be sure to specify the folder location when starting the development server by running `npx -y @adobe/aem-cli up --no-open --forward-browser-logs --html-folder drafts`
```

shsteimer committed:
Apply suggestion from @shsteimer

shsteimer committed:
Apply suggestion from @shsteimer

shsteimer committed:
Update AGENTS.md

added instructions to avoid bad/confusing css selector usage

fkakatie committed:
chore: fix typos for human readability

fkakatie commented on AGENTS.md:
```diff
@@ -0,0 +1,228 @@
+## Development Workflow
+
+### Local Development
+1. Run `npx -y @adobe/aem-cli up --no-open` to start the AEM Proxy server
```
Is this repetition from `## Setup Commands` wasting context?

fkakatie commented on AGENTS.md:
```diff
@@ -0,0 +1,228 @@
+
+## Common Tasks
+
+### Adding New Blocks
```
Necessary repetition from `### Block Development`?

fkakatie commented on AGENTS.md:
```diff
@@ -0,0 +1,228 @@
+- Follow the naming convention: `blockname.css`, `blockname.js`
+- Blocks should be responsive and accessible by default
+
+### Styling
```
Does this provide useful context beyond  `## Project Structure`?

fkakatie commented on AGENTS.md:
```diff
@@ -0,0 +1,228 @@
+1. Make changes to the specific block files
+2. Test locally with `aem up`
+3. Ensure responsive behavior across devices
+4. Run linting before committing
```
We've given linting instructions in multiple other places. Do we not trust an agent gets the memo?

royfielding reviewed (approved):
It seems odd that we are already linking to these files within the docs and using some version of them on the helix-website, but they are only on a branch for the boilerplate. I think it should be merged and the copies made consistent (if not already).

shsteimer reviewed (commented):
@fkakatie 

after some consideration, I think I'm in favor of merging this to main. If someone wants to delete it they can, but it causes no issues if you aren't using it, and I think the vast majority of repos at this point will likely be using some type of agentic coding tool. So I don't think having it on `main` violates the strict nothing you need to delete principle.

I also very much think we should consolidate it as much as possible and remove any thing that is duplicative, contradictory, or potentially confusing, per many of the suggestions made already.

There's probably an opportunity to use sub-agents and evals to be more systematic about how we do that, but for now, I'd say anything that seems duplicated or redundant is safe to remove.

rofe commented on AGENTS.md:
```diff
@@ -0,0 +1,228 @@
+        └── {blockName}.css     # Block's styles
+├── styles/          # Global styles and CSS
+    ├── styles.css          # Minimal global styling and layout for your website required for LCP
+    └── lazy-styles.css     # Additional global styling and layout for below the fold/post LCP content
```
Add `fonts.css`
```suggestion
    └── lazy-styles.css     # Additional global styling and layout for below the fold/post LCP content
    └── fonts.css           # Font definitions
```

shsteimer committed:
Update AGENTS.md

Co-authored-by: Raphael Wegmueller <github@rofe.com>

shsteimer committed:
fix: missing back tick for url in pr process

shsteimer committed:
cleanup: removed redundant info, clarified some things that agents may not know how to do, etc.

shsteimer committed:
add claude.md pointing to agents.md
